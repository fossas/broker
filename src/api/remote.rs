//! This module provides functionality for integrating with "remotes", which are external code hosts.
//!
//! # "Remote" as a concept
//!
//! An example of a [`Remote`] is Github; it's a place where code is stored remotely.
//! [`Remote`] as a concept is abstracted over protocol implementation:
//! for example, a code host accessed via `ssh` or via `http` are both simply a [`Remote`].
//!
//! Their connection protocol (and any specifics, like authentication) is then specified via
//! [`Protocol`], which is usually wrapped inside an [`Integration`], forming the primary interaction
//! point for this module.

use std::time::Duration;

use derive_more::{AsRef, Display, From};
use derive_new::new;
use error_stack::{report, IntoReport, Report, ResultExt};
use getset::{CopyGetters, Getters};
use humantime::parse_duration;
use url;

use crate::ext::error_stack::{DescribeContext, ErrorHelper};

pub mod git;

/// Errors that are possibly surfaced during validation of config values.
#[derive(Debug, thiserror::Error)]
pub enum ValidationError {
    /// Poll interval is parsed from a user-provided string.
    #[error("validate poll interval")]
    PollInterval,

    /// The provided remote is not valid.
    #[error("validate remote location")]
    Remote,

    /// The provided value is empty.
    #[error("value is empty")]
    ValueEmpty,
}

/// Validated config values for external code host integrations.
#[derive(Debug, Default, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct Config(Vec<Integration>);

/// Validated remote location for a code host.
#[derive(Debug, Clone, PartialEq, Eq, AsRef, Display, new)]
pub struct Remote(String);

impl Remote {
    /// parses a Remote and returns a Result<Url>
    pub fn parse(&self) -> Result<url::Url, url::ParseError> {
        url::Url::parse(&self.0)
    }
}

impl TryFrom<String> for Remote {
    type Error = Report<ValidationError>;

    fn try_from(input: String) -> Result<Self, Self::Error> {
        // Different remotes have different semantics-
        // we can't guarantee this is actually a well formatted URL.
        // Just validate that it's not empty.
        if input.is_empty() {
            Err(report!(ValidationError::ValueEmpty))
                .describe_lazy(|| format!("provided input: '{input}'"))
        } else {
            Ok(Remote(input))
        }
        .help("the remote location may not be empty")
        .change_context(ValidationError::Remote)
    }
}

/// Each integration has metadata (for example, how often it should be polled)
/// along with its protocol (which describes how to download the code so it can be analyzed).
///
/// This type stores this combination of data.
#[derive(Debug, Clone, PartialEq, Eq, Getters, CopyGetters, new)]
pub struct Integration {
    /// The interval at which Broker should poll the remote code host for whether the code has changed.
    #[getset(get_copy = "pub")]
    poll_interval: PollInterval,

    /// The protocol Broker uses to communicate with the remote code host.
    #[getset(get = "pub")]
    protocol: Protocol,
}

/// Code is stored in many kinds of locations, from git repos to
/// random FTP sites to DevOps hosts like GitHub.
///
/// To handle this variety, Broker uses a predefined list
/// of supported protocols (this type),
/// which are specialized with configuration unique to those integrations.
#[derive(Debug, Clone, PartialEq, Eq, From, new)]
pub enum Protocol {
    /// Integration with a code host using the git protocol.
    Git(git::Transport),
}

/// Specifies the maximum age for an observability artifact.
#[derive(Debug, Copy, Clone, PartialEq, Eq, AsRef, From, new)]
pub struct PollInterval(Duration);

impl TryFrom<String> for PollInterval {
    type Error = Report<ValidationError>;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        parse_duration(&value)
            .into_report()
            .change_context(ValidationError::PollInterval)
            .describe_lazy(|| format!("provided value: {value}"))
            .map(PollInterval)
    }
}
