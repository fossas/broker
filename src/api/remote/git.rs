pub mod repository;
pub mod transport;
use std::fmt::Display;

use derive_new::new;
use serde::{Deserialize, Serialize};

/// A git reference's type (branch or tag)
#[derive(Debug, Clone, Hash, Eq, PartialEq, Serialize, Deserialize, new)]
pub enum Reference {
    /// A branch
    Branch {
        /// The name of the branch
        name: String,

        /// The head commit of the branch
        head: String,
    },

    /// A tag
    Tag {
        /// The name of the tag
        name: String,
        /// The commit that the tag points at
        commit: String,
    },
}

impl Display for Reference {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Reference::Branch { name, head } => {
                write!(f, "branch::{name}@{head}")
            }
            Reference::Tag { name, commit } => {
                write!(f, "tag::{name}@{commit}")
            }
        }
    }
}

impl Reference {
    /// Retrieves name of reference
    pub fn name(&self) -> &String {
        match self {
            Self::Branch { name, .. } => name,
            Self::Tag { name, .. } => name,
        }
    }

    /// Generate a canonical state for the reference.
    pub fn as_state(&self) -> &[u8] {
        match self {
            Reference::Branch { head, .. } => head.as_bytes(),
            Reference::Tag { commit, .. } => commit.as_bytes(),
        }
    }

    /// Generate a representation for the reference suitable for use when
    /// creating database coordinates.
    pub fn for_coordinate(&self) -> String {
        match self {
            Reference::Branch { name, head } => format!("branch:{name}@{head}"),
            Reference::Tag { name, commit } => format!("tag:{name}@{commit}"),
        }
    }
}
