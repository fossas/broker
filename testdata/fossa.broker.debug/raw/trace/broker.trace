{"timestamp":"2023-04-05T16:30:24.086526Z","level":"INFO","fields":{"message":"Debug artifacts being stored in '/Users/me/.config/fossa/broker/debug/trace'"},"target":"broker::debug"}
