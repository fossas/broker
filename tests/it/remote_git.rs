//! Tests for git remotes
use crate::helper::assert_error_stack_snapshot;
use crate::{args::raw_base_args, helper::load_config};
use broker::api::remote::RemoteProvider;
use std::path::PathBuf;

use tempfile::tempdir;

use broker::{self, api::remote::git, config};

#[tokio::test]
async fn update_clones_on_public_repo_with_no_auth() {
    let conf = load_config!(
        "testdata/config/fossa-one-http-no-auth.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    let mut integrations = conf.integrations().as_ref().iter();
    let integration = integrations
        .next()
        .expect("no integration loaded from config");
    let tmpdir = tempdir().expect("creating tmpdir");
    let path = PathBuf::from(tmpdir.path());
    let mut expected_clone_paths: Vec<PathBuf> = vec!["master", "other-branch", "1.1"]
        .into_iter()
        .map(|reference| {
            let mut path = PathBuf::from(tmpdir.path());
            path.push(String::from(reference));
            path
        })
        .collect();
    expected_clone_paths.sort();
    let mut clone_paths = git::repository::Repository::update_clones(path, integration)
        .expect("no results returned from update_clones on a public repo!");
    clone_paths.sort();
    assert_eq!(expected_clone_paths, clone_paths);
}

#[tokio::test]
async fn clone_private_repo_with_no_auth() {
    let conf = load_config!(
        "testdata/config/private-repo-http-no-auth.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    let mut integrations = conf.integrations().as_ref().iter();
    let integration = integrations.next().unwrap();
    let tmpdir = tempdir().unwrap();
    let path = PathBuf::from(tmpdir.path());
    let context = String::from("cloning private repo with bad auth");
    assert_error_stack_snapshot!(
        &context,
        git::repository::Repository::update_clones(path, integration)
            .expect_err("Could not read from remote repository")
    );
}
