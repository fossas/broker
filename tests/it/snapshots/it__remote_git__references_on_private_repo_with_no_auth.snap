---
source: tests/it/remote_git.rs
expression: err
info: references on private repo with bad auth
---
run external command
├╴at {source location}
│
╰─▶ run command: git
    args: ["-c", "credential.helper=", "ls-remote", "--quiet", "http://github.com/github/doesnotexist.git"]
    env: ["GIT_TERMINAL_PROMPT='0'", "GCM_INTERACTIVE='never'", "GIT_ASKPASS=<REMOVED>"]
    status: 128
    stdout: ''
    stderr: 'fatal: could not read Username for 'https://github.com': terminal prompts disabled'
    ╰╴at {source location}
